arcs/src/algorithms/scale_non_uniform.rs

```diff
@@ -0,0 +1,101 @@
+use crate::{
+    primitives::{Point, Line},
+    Vector,
+};
+use kurbo::Affine;
+
+/// Something which can be scaled **non-uniform** in x and y directions in *Drawing Space*
+pub trait ScaleNonUniform {
+    fn scale_nu(&mut self, factor_x: f64, factor_y: f64, base: Vector);
+
+    fn scaled_nu(&self, factor_x: f64, factor_y: f64, base: Vector) -> Self
+    where 
+        Self: Sized + Clone,
+        {
+            let mut clone = self.clone();
+            clone.scale_nu(factor_x, factor_y, base);
+
+            clone
+        }
+
+}
+
+impl<'t, T: ScaleNonUniform + ?Sized> ScaleNonUniform for &'t mut T {
+    fn scale_nu(&mut self, factor_x: f64, factor_y: f64, base: Vector) {
+        (*self).scale_nu(factor_x, factor_y, base);
+    }
+}
+
+impl ScaleNonUniform for Vector {
```

Comment:
Are there any other types that could implement `ScaleNonUniform` and `Scale`? `BoundingBox` would be a good candidate.

It almost feels like we could use a more general `AffineTransformable` trait which accepts a `kurbo::Affine` to transform itself with. That way we could automatically implement `Scale`, `ScaleNonUniform`, `Translate`, and `Rotate` for any `T: AffineTransformable`... Thoughts?

Reply:
`BoundingBox` could also implement `Translate` then.

Using a `AffineTransformable` trait we could avoid writing all those seperate files.
So should we still use a layer of abstraction, e.g. a `Transformation` struct wich has methods for returning `Affine` instances?

We still would need to think about how to handle *Transformations*  which are possibly **destructive** e.g. `shear` and `non-uniform`. 

Maybe implementing a `Primitive` enum and a return an instance of it from the *transform* function when using *destuctive* transforms.

Reply:
> So should we still use a layer of abstraction, e.g. a Transformation struct wich has methods for returning Affine instances?

Isn't that just the various constructor functions on `kurbo::Affine`?

What is your definition of destructive? My thoughts are that if you can apply an affine transform and the end product is still the same type of shape, it's okay. 

So `Line` could implement `AffineTransformation` because stretching and scaling a `Line` still gives you a `Line`, but if you shear a `BoundingBox` it turns into a rhombus and is no longer an axis-aligned bounding box, so it wouldn't make sense for it to implement `AffineTransformable` and you'd need to implement `Scale`, `ScaleNonUniform`, and `Translate` manually. Same with `Arc`.

Reply:
Ok that makes a lot of sense, so we define one base trait `AffineTransformable` with all the transforms *auto-implemented* and if primitives need other transforms they get it implemented manually.

First draft for `AffineTransformable`:

```rust
pub trait AffineTransformable {
    fn transform(&mut self, affine: Affine, base: Vector);

    fn transformed(&self, affine: Affine, base: Vector) -> Self
    where 
        Self: Sized + Clone,
        {
            let mut clone = self.clone();
            clone.transform(affine, base);

            clone
        }

    fn translate(&mut self, displacement: Vector)
    where
        Self: Sized,
        {
            let affine = Affine::translate(displacement);
            self.transform(affine, Vector::zero());
        }

    fn translated(&mut self, displacement: Vector) -> Self
    where
        Self: Sized + Clone,
        {
            let mut clone = self.clone();
            clone.translate(displacement);

            clone
        }

    // TODO: implement other transforms, e.g. scale, scale-non-constanst, etc...
}
```

Reply:
I was thinking more along the lines of [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=111950e31908fc8b3b66329caffea682).

My thoughts are as follows:

- Make each transformation its own trait, so each type only implements the things that make sense for them (i.e. `translate()` doesn't belong in the `AffineTransformable` trait)
- Keep each trait [as small as possible](https://www.integralist.co.uk/posts/go-interfaces/#keep-interfaces-small), preferably with just one core method
- Where some transformation can be implemented *exactly* using a more general transformation (`Scale` is just a `ScaleNonUniform` where both factors are the same, `Translate` is just a `AffineTransformable` with an `Affine::translate()`) add a generic impl so you don't need to type out code twice (i.e. `impl <S: ScaleNonUniform> Scale for S { ... }`)
- Methods like `translated()` and `scaled()` which take an immutable reference and return a mutated copy are purely for convenience, so it's okay for them to have a `where Self: ...` clause and default implementation

Reply:
I will have time to really look at this starting tomorrow.

My first impressions:
  - Thank you for the detailed write-up
  - Keeping traits (*interfaces*) small sounds logical and good

I didn't know this pattern of writing traits and it looks like a real good foundation for the transformation system here